### ISSUE TITLE
Exception in 0.4.0-rc.1 - CacheManager._saveDataInPrefs toMap() was called in null

### ISSUE THREAD

**Issue by lemonboston** [renefloor/flutter_cached_network_image#25]
I just got this exception:
```
E/flutter (19933): [ERROR:topaz/lib/tonic/logging/dart_error.cc(16)] Unhandled exception:
E/flutter (19933): 'package:flutter/src/painting/image_stream.dart': Failed assertion: line 362: '_codec != null': is not true.
E/flutter (19933): #0      _AssertionError._doThrowNew (dart:core-patch/dart:core/errors_patch.dart:37)
E/flutter (19933): #1      _AssertionError._throwNew (dart:core-patch/dart:core/errors_patch.dart:33)
E/flutter (19933): #2      MultiFrameImageStreamCompleter._handleCodecReady (package:flutter/src/painting/image_stream.dart:362)
E/flutter (19933): #3      _RootZone.runUnary (dart:async/zone.dart:1381)
E/flutter (19933): #4      _FutureListener.handleValue (dart:async/future_impl.dart:129)
E/flutter (19933): #5      _Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:633)
E/flutter (19933): #6      _Future._propagateToListeners (dart:async/future_impl.dart:662)
E/flutter (19933): #7      _Future._completeError (dart:async/future_impl.dart:485)
E/flutter (19933): #8      _SyncCompleter._completeError (dart:async/future_impl.dart:55)
E/flutter (19933): #9      _Completer.completeError (dart:async/future_impl.dart:27)
E/flutter (19933): #10     CachedNetworkImageProvider._loadAsync (package:cached_network_image/cached_network_image.dart)
E/flutter (19933): <asynchronous suspension>
E/flutter (19933): #11     CachedNetworkImageProvider.load (package:cached_network_image/cached_network_image.dart:457)
E/flutter (19933): #12     ImageProvider.resolve.<anonymous closure>.<anonymous closure> (package:flutter/src/painting/image_provider.dart:266)
E/flutter (19933): #13     ImageCache.putIfAbsent (package:flutter/src/painting/image_cache.dart:82)
E/flutter (19933): #14     ImageProvider.resolve.<anonymous closure> (package:flutter/src/painting/image_provider.dart:266)
E/flutter (19933): #15     SynchronousFuture.then (package:flutter/src/foundation/synchronous_future.dart:38)
E/flutter (19933): #16     ImageProvider.resolve (package:flutter/src/painting/image_provider.dart:264)
E/flutter (19933): #17     _ImageProviderResolver.resolve (package:cached_network_image/cached_network_image.dart:193)
E/flutter (19933): #18     _CachedNetworkImageState._resolveImage (package:cached_network_image/cached_network_image.dart:279)
E/flutter (19933): #19     _CachedNetworkImageState.didChangeDependencies (package:cached_network_image/cached_network_image.dart:261)
E/flutter (19933): #20     StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:3739)
E/flutter (19933): #21     ComponentElement.mount (package:flutter/src/widgets/framework.dart:3600)
E/flutter (19933): #22     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2890)
E/flutter (19933): #23     MultiChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:4716)
E/flutter (19933): #24     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2890)
E/flutter (19933): #25     Element.updateChild (package:flutter/src/widgets/framework.dart:2693)
E/flutter (19933): #26     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3636)
E/flutter (19933): #27     Element.rebuild (package:flutter/src/widgets/framework.dart:3478)
E/flutter (19933): #28     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:3605)
E/flutter (19933): #29     ComponentElement.mount (package:flutter/src/widgets/framework.dart:3600)
E/flutter (19933): #30     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2890)
E/flutter (19933): #31     MultiChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:4716)
E/flutter (19933): #32     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2890)
E/flutter (19933): #33     Element.updateChild (package:flutter/src/widgets/framework.dart:2693)
E/flutter (19933): #34     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3636)
E/flutter (19933): #35     Element.rebuild (package:flutter/src/widgets/framework.dart:3478)
E/flutter (19933): #36     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:3605)
E/flutter (19933): #37     ComponentElement.mount (package:flutter/src/widgets/framework.dart:3600)
E/flutter (19933): #38     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2890)
E/flutter (19933): #39     Element.updateChild (package:flutter/src/widgets/framework.dart:2693)
E/flutter (19933): #40     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3636)
E/flutter (19933): #41     Element.rebuild (package:flutter/src/widgets/framework.dart:3478)
E/flutter (19933): #42     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:3605)
E/flutter (19933): #43     StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:3741)
E/flutter (19933): #44     ComponentElement.mount (package:flutter/src/widgets/framework.dart:3600)
E/flutter (19933): #45 
E/flutter (19933): [ERROR:topaz/lib/tonic/logging/dart_error.cc(16)] Unhandled exception:
E/flutter (19933): NoSuchMethodError: The method 'toMap' was called on null.
E/flutter (19933): Receiver: null
E/flutter (19933): Tried calling: toMap()
E/flutter (19933): #0      Object.noSuchMethod (dart:core-patch/dart:core/object_patch.dart:46)
E/flutter (19933): #1      CacheManager._saveDataInPrefs.<anonymous closure>.<anonymous closure> (package:flutter_cache_manager/flutter_cache_manager.dart:106)
E/flutter (19933): #2      _HashVMBase&MapMixin&&_LinkedHashMapMixin.forEach (dart:collection-patch/dart:collection/compact_hash.dart:361)
E/flutter (19933): #3      CacheManager._saveDataInPrefs.<anonymous closure> (package:flutter_cache_manager/flutter_cache_manager.dart:105)
E/flutter (19933): #4      SynchronizedLock._runTask.<anonymous closure>.<anonymous closure> (package:synchronized/src/synchronized_impl.dart:194)
E/flutter (19933): #5      _rootRun (dart:async/zone.dart:1126)
E/flutter (19933): #6      _CustomZone.run (dart:async/zone.dart:1023)
E/flutter (19933): #7      runZoned (dart:async/zone.dart:1501)
E/flutter (19933): #8      SynchronizedLock._runTask.<anonymous closure> (package:synchronized/src/synchronized_impl.dart:192)
E/flutter (19933): #9      new Future.sync (dart:async/future.dart:222)
E/flutter (19933): #10     SynchronizedLock._runTask (package:synchronized/src/synchronized_impl.dart:191)
E/flutter (19933): #11     LockBase._createAndRunTask.run (package:synchronized/src/synchronized_impl.dart:66)
E/flutter (19933): #12     LockBase._createAndRunTask (package:synchronized/src/synchronized_impl.dart:76)
E/flutter (19933): #13     SynchronizedLock.synchronized (package:synchronized/src/synchronized_impl.dart:227)
E/flutter (19933): #14     synchronized (package:synchronized/src/synchronized_impl.dart:280)
E/flutter (19933): #15     synchronized (package:synchronized/synchronized.dart:47)
E/flutter (19933): #16     CacheManager._saveDataInPrefs (package:flutter_cache_manager/flutter_cache_manager.dart:104)
E/flutter (19933): <asynchronous suspension>
E/flutter (19933): #17     CacheManager._save (package:flutter_cache_manager/flutter_cache_manager.dart:76)
E/flutter (19933): <asynchronous suspension>
E/flutter (19933): #18     CacheManager.getFile (package:flutter_cache_manager/flutter_cache_manager.dart:237)
E/flutter (19933): <asynchronous suspension>
E/flutter (19933): #19     CachedNetworkImageProvider._loadAsync (package:cached_network_image/cached_network_image.dart:469)
E/flutter (19933): <asynchronous suspension>
E/flutter (19933): #20     CachedNetworkImageProvider.load (package:cached_network_image/cached_network_image.dart:457)
E/flutter (19933): #21     ImageProvider.resolve.<anonymous closure>.<anonymous closure> (package:flutter/src/painting/image_provider.dart:266)
E/flutter (19933): #22     ImageCache.putIfAbsent (package:flutter/src/painting/image_cache.dart:82)
E/flutter (19933): #23     ImageProvider.resolve.<anonymous closure> (package:flutter/src/painting/image_provider.dart:266)
E/flutter (19933): #24     SynchronousFuture.then (package:flutter/src/foundation/synchronous_future.dart:38)
E/flutter (19933): #25     ImageProvider.resolve (package:flutter/src/painting/image_provider.dart:264)
E/flutter (19933): #26     _ImageProviderResolver.resolve (package:cached_network_image/cached_network_image.dart:193)
E/flutter (19933): #27     _CachedNetworkImageState._resolveImage (package:cached_network_image/cached_network_image.dart:279)
E/flutter (19933): #28     _CachedNetworkImageState.didChangeDependencies (package:cached_network_image/cached_network_image.dart:261)
E/flutter (19933): #29     StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:3739)
E/flutter (19933): #30     ComponentElement.mount (package:flutter/src/widgets/framework.dart:3600)
E/flutter (19933): #31     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2890)
E/flutter (19933): #32     MultiChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:4716)
E/flutter (19933): #33     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2890)
E/flutter (19933): #34     Element.updateChild (package:flutter/src/widgets/framework.dart:2693)
E/flutter (19933): #35     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3636)
E/flutter (19933): #36     Element.rebuild (package:flutter/src/widgets/framework.dart:3478)
E/flutter (19933): #37     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:3605)
E/flutter (19933): #38     ComponentElement.mount (package:flutter/src/widgets/framework.dart:3600)
E/flutter (19933): #39     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2890)
E/flutter (19933): #40     MultiChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:4716)
E/flutter (19933): #41     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2890)
E/flutter (19933): #42     Element.up
```

**Comment by renefloor** [comment]
`_codec != null` should only be called when the image was not cached and it failed to load. Could you confirm that? Maybe you weren't connected to the internet or something like that? I have not been able to catch that error yet, but it shouldn't give a problem for the user.
`_saveDataInPrefs: The method 'toMap' was called on null.` should never happen, but it probably does when the first image fails to load and thus there is not data yet to save. I will fix that.

**Comment by lemonboston** [comment]
Yes, I can confirm that this happened indeed when I was testing it offline. And it didn't cause any problem on the UI (just stopped the app running in debug mode at the exception).
Feel free of course to close this as you see fit.

By the way, I've successfully replaced my image widgets with `CachedNetworkImage` and everything works nicely as expected. So thanks for the library! I helps our project a lot now.

**Comment by renefloor** [comment]
The error in `_saveDataInPrefs` should definitely be fixed, so I keep this issue open. I'm happy to have helped you with this project :)

**Comment by renefloor** [comment]
This should be fixed now in 0.4.0
